wasm.js

```diff
@@ -0,0 +1,185 @@
+/*
+ * This file is part of the 0chain @zerochain/0chain distribution (https://github.com/0chain/client-sdk).
+ * Copyright (c) 2018 0chain LLC.
+ *
+ * 0chain @zerochain/0chain program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+'use strict';
+
+const utils = require('./utils');
+
+// Initially, the __zcn_wasm__ object will be
+const g = global || window || self;
+
+/**
+ * bridge is an easier way to refer to the Go WASM object.
+ */
+const bridge = g.__zcn_wasm__;
+
+async function blsSign (hash) {
+  if (!bridge.bls?.secretKey) {
+    const msg = 'err: bls.secretKey is not initialized';
+    console.error(msg);
+    return msg;
+  }
+
+  try {
+    const bytes = utils.hexStringToByte(hash);
+
+    const sig = bridge.bls.secretKey.sign(bytes);
+
+    if (!sig) {
+      const msg = 'err: failed to sign';
+      return msg;
+    }
+
+    return sig.serializeToHexStr();
+  } catch (error) {
+    console.error(error);
+    return 'err: ' + error.Error();
+  }
+}
+
+if (!g.__zcn_wasm__) {
+	g.__zcn_wasm__ = {
+		bls: {
+			sign: blsSign,
+		},
+		sdk: {},
+	};
+}
+
+/**
+ * Wrapper is used by Go to run all Go functions in JS.
+ *
+ * @param {Function} goFunc a function that is expected to return an object of the following specification:
+ * {
+ *  result:  undefined | any         // undefined when error is returned, or function returns undefined.
+ *  error:       Error | undefined   // undefined when no error is present.
+ * }
+ *
+ * @returns {Function} returns a function that take arguments which are used to call the Go function.
+ */
+const wrapper = (goFunc) => (...args) => {
+	const response = goFunc.apply(undefined, args);
+
+	if (response.error instanceof Error) {
+		throw response.error;
+	}
```

Comment:
Do we need to check if the error is an `instanceof` Error? Even if it's only defined, it should be thrown.

```suggestion
	if (response.error instanceof Error)
	  throw response.error;
```

Reply:
`error` is not supported from go response.  need to double check it.

Reply:
```
// js.ValueOf returns x as a JavaScript value:
//
//  | Go                     | JavaScript             |
//  | ---------------------- | ---------------------- |
//  | js.Value               | [its value]            |
//  | js.Func                | function               |
//  | nil                    | null                   |
//  | bool                   | boolean                |
//  | integers and floats    | number                 |
//  | string                 | string                 |
//  | []interface{}          | new array              |
//  | map[string]interface{} | new object             |
//
// Panics if x is not one of the expected types.
```

Reply:
an object {error:""} as Error. BTW. wasm.js is designed for go. js features are not supported fully here. that is why I don't leave it with your team.

Reply:
The response comes from Go. But the handling of it has nothing to do with Go. This is pure javascript, no mystery about it.

Reply:
The error will be thrown for the client, after the Go function has run.

Reply:
in our wasm sdk , {error:""} will be returned to js if any exception is thrown, if not , result is returned to js directly. please check wasmsdk too if you want to understand whole logic.

Reply:
let's leave it with your team in future improvement if there is better way for your development